pkg/apis/core/types.go

```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies the node's canonical hostname
+	NodeHostName NodeAddressType = "Hostname"
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may or may not be visible to hosts outside the cluster. By default it is
+	// assumed that kube-apiserver can reach node internal IPs, though it is possible
+	// to configure clusters where this is not the case.
+	//
+	// In cloud environments, a NodeInternalIP address generally corresponds to an
+	// "internal" or "private" IP address which was automatically assigned to the
+	// node. In clusters with no cloud provider, NodeInternalIP is simply the standard
+	// address type, and does not necessarily imply that the IP is internal-only.
+	NodeInternalIP NodeAddressType = "InternalIP"
+
+	// NodeExternalIP identifies an IP address explicitly intended for use from
+	// outside the cluster; it is assumed to be more public or more persistent than a
+	// NodeInternalIP. (In public clouds, NodeExternalIP addresses are normally
```

Comment:
"NodeExternalPs can be globally routable unicast IPs."

Reply:
I was trying to be more specific/normative than that. Public clouds generally have globally routable IPs at their disposable, and generally have APIs for assigning them to nodes, and are generally used in contexts in which the users want some or all nodes to be reachable from across the internet. Private clouds come with a different set of assumptions. And all of the "classic" public CloudProviders (AWS, GCP, Azure) implement NodeExternalIP in the way described here, while none of the classic private CloudProviders do. So it seems to make sense to call this out as being an expected behavior specifically of "public cloud" CloudProviders.

Reply:
> And all of the "classic" public CloudProviders (AWS, GCP, Azure) implement NodeExternalIP in the way described here, while none of the classic private CloudProviders do.

On vSphere, a VM can have a publically routable IP external IP. It's not common but you can.